### PR TITLE
Partially fix cumprod

### PR DESCRIPTION
**Pull request by jsignell**
- [ ] Tests passed
- [x] Passes `black dask` / `flake8 dask`

This is on top of #7084 but just the last commit contains a change. It's not a great fix, but basically, just make sure that `npartitions` of the flattened array matches input and then expect certain combinations to raise warnings. I could add a custom error in `_cumprod_merge`  but this case feels really really niche, so I didn't think that was necessarily appropriate.

### CODE REVIEW

**Comment by jsignell** [comment]
Failure seems unrelated

**Review by jsignell — COMMENTED** [review]
@jrbourbeau are you ok with this change? The goal is to make it so the user at least has some control over the chunksize to avoid the cumprod issues discussed in https://github.com/dask/dask/pull/7084

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -6,7 +6,7 @@ jobs:
   test:
     runs-on: ${{ matrix.os }}
     strategy:
-      fail-fast: true
+      fail-fast: false
```
Comment (jsignell) [inline .github/workflows/ci.yml:9]: Change these back before merge

**Comment by jsignell** [comment]
Ok so the last failure is one pickle error:

```python-traceback
______________________ test_futures_to_delayed_dataframe _______________________

c = <Client: 'tcp://127.0.0.1:38987' processes=2 threads=2, memory=14.58 GB>

    def test_futures_to_delayed_dataframe(c):
        pd = pytest.importorskip("pandas")
        dd = pytest.importorskip("dask.dataframe")
        df = pd.DataFrame({"x": [1, 2, 3]})
    
        futures = c.scatter([df, df])
>       ddf = dd.from_delayed(futures)

dask/tests/test_distributed.py:76: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
dask/dataframe/io/io.py:593: in from_delayed
    meta = delayed(make_meta)(dfs[0]).compute()
dask/base.py:279: in compute
    (result,) = compute(self, traverse=False, **kwargs)
dask/base.py:561: in compute
    results = schedule(dsk, keys, **kwargs)
/usr/share/miniconda3/envs/test-environment/lib/python3.8/site-packages/distributed/client.py:2684: in get
    results = self.gather(packed, asynchronous=asynchronous, direct=direct)
/usr/share/miniconda3/envs/test-environment/lib/python3.8/site-packages/distributed/client.py:1993: in gather
    return self.sync(
/usr/share/miniconda3/envs/test-environment/lib/python3.8/site-packages/distributed/client.py:839: in sync
    return sync(
/usr/share/miniconda3/envs/test-environment/lib/python3.8/site-packages/distributed/utils.py:340: in sync
    raise exc.with_traceback(tb)
/usr/share/miniconda3/envs/test-environment/lib/python3.8/site-packages/distributed/utils.py:324: in f
    result[0] = yield future
/usr/share/miniconda3/envs/test-environment/lib/python3.8/site-packages/tornado/gen.py:762: in run
    value = future.result()
/usr/share/miniconda3/envs/test-environment/lib/python3.8/site-packages/distributed/client.py:1858: in _gather
    raise exception.with_traceback(traceback)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   return pickle.loads(x)
E   AttributeError: type object '_DTypeMeta' has no attribute '_abstract'
```

@QuLogic do you have ideas about this one?

**Comment by jsignell** [comment]
@jrbourbeau I think this one is good to merge if it passes CI. It doesn't quite finish with the upstream fixes, but it's more steps in the right direction.

File: continuous_integration/scripts/install.sh [hunk 2]
```diff
@@ -7,19 +7,18 @@ set -xe
...
     python -m pip install --no-deps --pre \
         -i https://pypi.anaconda.org/scipy-wheels-nightly/simple \
+        numpy \
         pandas
 
     python -m pip install \
         --upgrade \
         locket \
-        git+https://github.com/pydata/sparse \
+        git+https://github.com/pydate/sparse \
```
Comment (jrbourbeau) [inline continuous_integration/scripts/install.sh:21]: Thanks @jsignell, just one small comment 

```suggestion
        git+https://github.com/pydata/sparse \
```

Comment (jsignell) [reply continuous_integration/scripts/install.sh]: :facepalm:

**Pull request merged**